livre/core/pipeline/FilterOutput.cpp

```diff
@@ -0,0 +1,223 @@
+    {
+        if( _portMap.count( portName ) > 0 )
+            return true;
+
+        return false;
+    }
+
+    void flush()
+    {
+        BOOST_FOREACH( const NameOutputPortPair& namePortPair,
+                       _portMap )
+        {
+            OutputPortPtr port = namePortPair.second;
+            port->flush();
+        }
+    }
+
+    bool waitForAny( lunchbox::Strings& readyPorts ) const
+    {
+        ConstOutputPortPtrs notReadyPorts;
+        BOOST_FOREACH( const NameOutputPortPair& namePortPair,
+                       _portMap )
+        {
+            const ConstOutputPortPtr& port = namePortPair.second;
+            const std::string& portName = port->getName( );
+            if( isReady( portName ))
+                readyPorts.push_back( portName );
+            else
+                notReadyPorts.push_back( port );
+        }
+
+        if( notReadyPorts.empty( ))
+            return false;
+
+        livre::waitForAny( notReadyPorts );
+        return true;
+    }
+
+    void wait( const std::string& portName = "" ) const
+    {
+        if( !portName.empty())
+        {
+            getPort( portName )->wait();
+            return;
+        }
+
+        BOOST_FOREACH( const NameOutputPortPair& namePortPair,
+                       _portMap )
+        {
+            namePortPair.second->wait();
+        }
+    }
+
+    bool isReady( const std::string& portName ) const
+    {
+        if( !portName.empty())
+            return getPort( portName )->isReady();
+
+        BOOST_FOREACH( const NameOutputPortPair& namePortPair,
+                       _portMap )
+        {
+            const ConstOutputPortPtr& port = namePortPair.second;
+            if( !port->isReady( ))
+                return false;
+        }
+
+        return true;
+    }
+
+    OutputPortMap _portMap;
+};
+
+FilterOutput::FilterOutput( )
+    : _impl( new FilterOutput::Impl())
+{}
+
+FilterOutput::~FilterOutput()
+{
+    
+}
+
+void FilterOutput::flush()
+{
+    _impl->flush();
+}
+
+OutputPortPtr FilterOutput::getPort( const std::string& portName )
+{
+    return _impl->getPort( portName  );
+}
+
+OutputPortPtr FilterOutput::addPort( const PortInfo& portInfo )
+{
+    return _impl->addPort( portInfo );
+}
+
+ConstOutputPortPtrs FilterOutput::getPorts() const
+{
+    return _impl->getPorts();
+}
+
+ConstOutputPortPtr FilterOutput::getPort( const std::string& portName ) const
+{
+    return _impl->getPort( portName );
+}
+
+bool FilterOutput::hasPort( const std::string& portName ) const
+{
+    return _impl->hasPort( portName );
+}
+
+Strings FilterOutput::getPortNames() const
+{
+    return _impl->getPortNames();
+}
+
+bool FilterOutput::waitForAny( Strings& readyPorts ) const
+{
+    return _impl->waitForAny( readyPorts );
+}
+
+void FilterOutput::wait( const std::string& portName ) const
+{
+    return _impl->wait( portName );
+}
+
+bool FilterOutput::isReady( const std::string& portName ) const
+{
+    return _impl->isReady( portName );
+}
+
+}
```

Comment:
I thought about it. At the beginning it felt they are really different and I did not access "Port" abstract object. So, I kept them seperate. I will think about it.